// cmd-ts-too/dist/lib/cmd-ts-too/batteries/fs.js
import {
  extendType,
  string
} from "../chunks/chunk-NZGMVZHM.js";

// src/batteries/fs.ts
import fs from "node:fs";
import path from "node:path";
var ExistingPath = extendType(string, {
  displayName: "path",
  description: "An existing path",
  async from(str) {
    const resolved = path.resolve(str);
    if (!fs.existsSync(resolved)) {
      throw new Error("Path doesn't exist");
    }
    return resolved;
  }
});
var Directory = extendType(ExistingPath, {
  async from(resolved) {
    const stat = fs.statSync(resolved);
    if (stat.isDirectory()) {
      return resolved;
    }
    return path.dirname(resolved);
  },
  displayName: "dir",
  description: "A path to a directory or a file within a directory"
});
var File = extendType(ExistingPath, {
  async from(resolved) {
    const stat = fs.statSync(resolved);
    if (stat.isFile()) {
      return resolved;
    }
    throw new Error("Provided path is not a file");
  },
  displayName: "file",
  description: "A file in the file system"
});
export {
  Directory,
  ExistingPath,
  File
};
